lib/datadog/security/contrib/rack/patcher.rb

```diff
@@ -0,0 +1,99 @@
+require 'datadog/security/contrib/patcher'
+require 'datadog/security/contrib/rack/integration'
+
+module Datadog
+  module Security
+    module Contrib
+      module Rack
+        module Patcher
+          include Datadog::Security::Contrib::Patcher
+
+          module_function
+
+          def patched?
+            Patcher.instance_variable_get(:@patched)
+          end
+
+          def target_version
+            Integration.version
+          end
+
+          # TODO: logger
+          def logger
+            return @logger if @logger
+
+            @logger ||= ::Logger.new(STDOUT)
+            #@logger.level = ::Logger::DEBUG
+            @logger.level = ::Logger::DEBUG
+            @logger.debug { 'logger enabled' }
+            @logger
+          end
+
+          def patch
+            Instrumentation.gateway.watch('rack.request') do |request|
+              block = false
+
+              Reactive::Operation.new('rack.request') do |op|
+                if defined?(Datadog::Tracer) && Datadog.respond_to?(:tracer) && (tracer = Datadog.tracer)
+                  root_span = Datadog.tracer.active_root_span
+                  active_span = Datadog.tracer.active_span
+
+                  logger.debug { "root span: #{root_span.span_id}"} if root_span
+                  logger.debug { "active span: #{active_span.span_id}"} if active_span
+                end
+
+                addresses = [
+                  'request.user_agent',
+                  'request.params',
+                  'request.headers',
+                  'request.referer',
+                  'request.path',
+                ]
+                op.subscribe(*addresses) do |*values|
```

Comment:
Is it necessary to set up the subscription on every single request? Could we do this only once at `patch` time?

Reply:
> Could we do this only once at patch time?

I think we could, although it requires storing the "structure" of the chain of events somehow, hoisting the creation of it out of the app runtime path, for it to be applied later at runtime. This would need a more involved design so I'd rather not veer into premature optimisation and produce a complex (and possibly unfit) design.

Reply:
> I wonder if we could go one level further and even create a struct to keep these values.

Yes, this should be better, all the more because it's quite obtuse this way (keeping track of the addresses vs values passed as args is only easy when both are very local to each other), which a struct would clarify, and also because type checking really doesn't like these very dynamic arguments, which a struct would solve.